Apps/W1/Shopify/app/src/Invoicing/Codeunits/ShpfyDraftOrdersAPI.Codeunit.al

```diff
@@ -0,0 +1,385 @@
+        DraftOrderId := SendDraftOrderGraphQLRequest(GraphQuery);
+        exit(DraftOrderId);
+    end;
+
+    /// <summary>
+    /// Completes a draft order in shopify by converting it to an order.
+    /// </summary>
+    /// <param name="DraftOrderId">Draft order id that needs to be completed.</param>
+    /// <returns>Json response of a created order in shopify.</returns>
+    internal procedure CompleteDraftOrder(DraftOrderId: BigInteger): JsonToken
+    var
+        GraphQLType: Enum "Shpfy GraphQL Type";
+        Parameters: Dictionary of [Text, Text];
+        JResponse: JsonToken;
+    begin
+        GraphQLType := "Shpfy GraphQL Type"::DraftOrderComplete;
+        Parameters.Add('DraftOrderId', Format(DraftOrderId));
+        JResponse := ShpfyCommunicationMgt.ExecuteGraphQL(GraphQLType, Parameters);
+        exit(JResponse);
+    end;
+
+    /// <summary>
+    /// Sets a global shopify shop to be used for draft orders api functionality.
+    /// </summary>
+    /// <param name="ShopCode">Shopify shop code to be set.</param>
+    internal procedure SetShop(ShopCode: Code[20])
+    begin
+        Clear(ShpfyShop);
+        ShpfyShop.Get(ShopCode);
+        ShpfyCommunicationMgt.SetShop(ShpfyShop);
+    end;
+
+    local procedure CreateDraftOrderGQLRequest(
+        var TempShpfyOrderHeader: Record "Shpfy Order Header" temporary;
+        var TempShpfyOrderLine: Record "Shpfy Order Line" temporary;
+        var ShpfyOrderTaxLines: Dictionary of [Text, Decimal]
+    ): TextBuilder
+    var
+        GraphQuery: TextBuilder;
+    begin
+        GraphQuery.Append('{"query":"mutation {draftOrderCreate(input: {');
+        if TempShpfyOrderHeader.Email <> '' then begin
+            GraphQuery.Append('email: \"');
+            GraphQuery.Append(ShpfyCommunicationMgt.EscapeGrapQLData(TempShpfyOrderHeader.Email));
+            GraphQuery.Append('\"');
+        end;
+        if TempShpfyOrderHeader."Phone No." <> '' then begin
+            GraphQuery.Append('phone: \"');
+            GraphQuery.Append(ShpfyCommunicationMgt.EscapeGrapQLData(TempShpfyOrderHeader."Phone No."));
+            GraphQuery.Append('\"');
+        end;
+        if TempShpfyOrderHeader."Currency Code" <> '' then begin
+            GraphQuery.Append('presentmentCurrencyCode: ');
+            GraphQuery.Append(Format(GetISOCode(TempShpfyOrderHeader."Currency Code")));
+        end;
+        if TempShpfyOrderHeader."Discount Amount" <> 0 then
+            AddDiscountAmountToGraphQuery(GraphQuery, TempShpfyOrderHeader."Discount Amount", 'Invoice Discount Amount');
+
+        GraphQuery.Append(', taxExempt: true');
+
+        AddShippingAddressToGraphQuery(GraphQuery, TempShpfyOrderHeader);
+        AddBillingAddressToGraphQuery(GraphQuery, TempShpfyOrderHeader);
+        AddLineItemsToGraphQuery(GraphQuery, TempShpfyOrderHeader, TempShpfyOrderLine, ShpfyOrderTaxLines);
```

Comment:
Unless order matters, I'd move lines to the end. It feels a bit odd, you set up all header info, then lines, and then again header a bit.